### ISSUE TITLE
Issues with native compilation when using straight.el and use-package

### ISSUE THREAD

**Issue by Panadestein** [joaotavora/yasnippet#1166]
Hello,

I'm having an issue with `yasnippet` when using `straight.el` and `use-package`.

Here's my setup:

   Emacs version: 3a50773ab0071addba98249d26c309f5fb78bd74
    yasnippet version: master
    Package setup: Installed via straight.el and use-package

Here's the configuration I use to set up yasnippet:

```emacs-lisp
(use-package yasnippet
  :straight t
  :config
  (yas-global-mode 1))
```

When I run `emacs --debug-init` I get the following error:

```emacs-lisp
⛔ Warning (initialization): An error occurred while loading ‘/home/loren/.emacs.d/init.el’:

Symbol's function definition is void: (setf internal--with-suppressed-warnings)

To ensure normal operation, you should investigate and remove the
cause of the error in your initialization file.  Start Emacs with
the ‘--debug-init’ option to view a complete error backtrace.
```

If I run some command afterward, I get this error:

```emacs-lisp
yas--template-menu-binding-pair-get-create: Symbol's function definition is void: \(setf\ internal--with-suppressed-warnings\)
```

How can this be solved? Thanks in advance.

**Comment by SjB** [comment]
I'm also seeing this error. on GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8) of 2023-06-25 using the Melpa 20200604.246 version of  yasnippets.

**Comment by SjB** [comment]
The error is from the [yasnippet-snippets](https://github.com/AndreaCrotti/yasnippet-snippets) package.

**Comment by Panadestein** [comment]
@SjB how did you debug this? Did you find a workaround?

**Comment by SjB** [comment]
No workaround yet didn't have time.

**Comment by dalugm** [comment]
> The error is from the [yasnippet-snippets](https://github.com/AndreaCrotti/yasnippet-snippets) package.

I guess not, I'm not using `yasnippet-snippets` and I still met this error.

And this error also happens in my config, one in `borg` managed, one in `use-package` managed.

I think it's the new version Emacs leads to this.

Using Emacs 29.0.92 works fine, but using the master branch broke.

The only problem is to figure out which commit leads this...

**Comment by Panadestein** [comment]
There is relevant information on the Emacs [mailing list](https://lists.gnu.org/archive/html/emacs-devel/2023-06/msg00776.html). It has been suggested that it is related to macroexpand.

**Comment by Panadestein** [comment]
I haven't debugged in detail yet, but rolling back to 8806bbbf2cdae00faacab9cec634072ce4e80b6f lets me use yasnippet again.

**Comment by dalugm** [comment]
This issue seems like related to this:

https://debbugs.gnu.org/cgi/bugreport.cgi?bug=64292

After add `lexical-binding: t` in `yasnippets.el`, I can use yasnippet just like before.

**Comment by edkolev** [comment]
Adding `lexical-binding: t` in yasnippets.el mitigates the issue for me as well. However, I can't help but wonder what it breaks in yasnippet :)

**Comment by Panadestein** [comment]
Shall we make a pull request adding it?

**Comment by edkolev** [comment]
I don't think it's as simple as just enabling it - there might be places in the code which rely on a dynamic (as opposed to lexical) scoping. The challenge would be to check there are no regressions after the switch.

I see this comment mentions making the switch https://github.com/joaotavora/yasnippet/issues/926#issuecomment-380424668

**Comment by Panadestein** [comment]
As discussed in the [emacs-devel](https://lists.gnu.org/archive/html/emacs-devel/2023-06/msg00815.html) mailing list, the issue got solved in dcd92090950929eeea3040fcac0294b7a4d839f0.
